src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs

```diff
@@ -703,5 +689,39 @@ public static void CannotSet_DefaultIgnoreCondition_To_Always()
         {
             Assert.Throws<ArgumentException>(() => new JsonSerializerOptions { DefaultIgnoreCondition = JsonIgnoreCondition.Always });
         }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/36605")]
+        public static void ConverterRead_VerifyInvalidTypeToConvertFails()
+        {
+            var options = new JsonSerializerOptions();
+            Type typeToConvert = typeof(KeyValuePair<int, int>);
+            byte[] bytes = Encoding.UTF8.GetBytes(@"{""Key"":1,""Value"":2}");
+
+            JsonConverter<KeyValuePair<int, int>> converter =
+                (JsonConverter<KeyValuePair<int, int>>)options.GetConverter(typeToConvert);
+
+            // Baseline
+            var reader = new Utf8JsonReader(bytes);
+            reader.Read();
+            KeyValuePair<int, int> kvp = converter.Read(ref reader, typeToConvert, options);
+            Assert.Equal(1, kvp.Key);
+            Assert.Equal(2, kvp.Value);
+
+            // Test
+            reader = new Utf8JsonReader(bytes);
+            reader.Read();
+            try
+            {
+                converter.Read(ref reader, typeof(Dictionary<string, int>), options);
```

Comment:
So this should throw InvalidOperation but today we throw NullRef?

Reply:
Yes, I think it should throw InvalidOperation.

I don't think we'll null ref in all cases. I just observed one in the converter for parameterized ctors due to the converter being passed a type that it can't convert (due to wrong assumptions about metadata on `state.Current.*` ). We may have the same/other issues with other converters.